crates/ruff_linter/src/checkers/ast/analyze/statement.rs

```diff
@@ -19,6 +19,9 @@ use crate::settings::types::PythonVersion;
 
 /// Run lint rules over a [`Stmt`] syntax node.
 pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
+    if checker.settings.rules.enabled(Rule::UnsortedDunderAll) {
+        ruff::rules::sort_dunder_all(checker, stmt);
+    }
```

Comment:
I'd recommend adding separate methods in `sort_dunder_all` for `Assign` and `AugAssign` (and probably also `AnnAssign`?), and then have those call into a single method internally that takes the statement and value. That way, we ensure we're only calling this rule on the correct AST nodes.

Reply:
I wondered what you might say about this bit :)

Just to double-check I understand: I should have three separate functions in `sort_dunder_all.rs`, with these signatures?

```rs
pub(crate) fn sort_dunder_all_assign(checker: &mut Checker, stmt: &ast::StmtAssign) {}
pub(crate) fn sort_dunder_all_annassign(checker: &mut Checker, stmt: &ast::StmtAnnAssign) {}
pub(crate) fn sort_dunder_all_assign(checker: &mut Checker, stmt: &ast::StmtAugAssign) {}
```

And then call out to those from different places in `checkers/ast/analyze/statement.rs`?

Reply:
@charliermarsh Is this how the existing lints work? It is interesting to me because it seems nice for the lint implementation itself to own when it is applicable. Otherwise, you wind up redoing the case analysis. Although I suppose if _all_ lints did this, then every lint would do its own case analysis and maybe that would be bad for perf reasons. I think with your suggested approach, maybe most lints don't need to do any case analysis.

@AlexWaygood Your code snippet looks right to me. Although I do wonder if you could avoid re-doing case analysis on `ast::Stmt`. Basically, have your internal function accept a `target` and `original_value`. Not 100% sure on that though.

Reply:
> Although I do wonder if you could avoid re-doing case analysis on `ast::Stmt`. Basically, have your internal function accept a `target` and `original_value`. Not 100% sure on that though.

Charlie [called me out on that](https://github.com/astral-sh/ruff/pull/9313#discussion_r1438924016) last time I tried to do something like that 😅

Reply:
Right. I don't mean that. I'm still proposing that you have the three functions @charliermarsh suggested above as the _public_ API for your new lint module. But internally, instead of re-materializing an `ast::Stmt` and just pattern matching on it again, you extract the state you need and pass it to your main internal routine that implements the lint.

Reply:
Ah, I think we're on the same page! I just pushed cc42bf9 -- is that what you had in mind?

Reply:
Precisely. :-)